include/net/fota_download.h

```diff
@@ -57,6 +57,8 @@ enum fota_download_error_cause {
 	FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE,
 	/** Actual firmware type does not match expected. Retry will not help. */
 	FOTA_DOWNLOAD_ERROR_CAUSE_TYPE_MISMATCH,
+	/** Downloading the update failed. The download may be retried. */
+	FOTA_DOWNLOAD_ERROR_CAUSE_CONNECTION_LOST,
```

Comment:
connection lost means that during download there were some connection issues (no network, network issues causing too many retries, socket issues (dl server does not respond). download failed is more caused by configuration/internal issues (invalid packages received etc). 

For an administrator this information is quite useful for troubleshooting. I hope you see my point :)

Reply:
Well, not really.
You seem to be trying to create distinction between download failures and network failures. In reality, this might not be possible. Download libraries only use Socket API, and they are not following the state of any network interfaces.

I would personally prefer to keep this API generic and simple. So if the outcome of the `CONNECTION_LOST` and `DOWNLOAD_FAILED` is exactly the same. You react exactly same, then those error codes should be the same.

Reply:
Take a look at this routine you find in zephyr/subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c. I understand you point to make the interface simple. The code you see below illustrates why I still would like to distinguish between `CONNECTION_LOST` and `DOWNLOAD_FAILED`. For the download client this is valid for the function `coap_initiate_retransmission` you find in `sdk-nrf/subsys/net/lib/download_client/src/coap.c` that limits the number of retry messages. For me this is a big difference between if we stop the download manually or just an error occurs during the download. If e.g. some retry count exceeds the user can actively fix this issue by improving the connection (moving the device to a different location) or simply incrementing the retry count. With my background of supplying thousands of device with a fw update for many years, it was always important for me to know as much as I can why an update failed. Sending just a `DOWNLOAD_FAILED` is somewhat of generic and does not show the reason for a failing update.

```
static void do_transmit_timeout_cb(struct lwm2m_message *msg)
{
    int ret;

    if (firmware_retry < PACKET_TRANSFER_RETRY_MAX) {
        /* retry block */
        LOG_WRN("TIMEOUT - Sending a retry packet!");

        ret = transfer_request(&firmware_block_ctx,
                       msg->token, msg->tkl,
                       do_firmware_transfer_reply_cb);
        if (ret < 0) {
            /* abort retries / transfer */
            set_update_result_from_error(ret);
            firmware_retry = PACKET_TRANSFER_RETRY_MAX;
            lwm2m_engine_context_close(&firmware_ctx);
            return;
        }

        firmware_retry++;
    } else {
        LOG_ERR("TIMEOUT - Too many retry packet attempts! "
            "Aborting firmware download.");
        lwm2m_firmware_set_update_result(RESULT_CONNECTION_LOST);
        lwm2m_engine_context_close(&firmware_ctx);
    }
}
```

Reply:
OK.
I think that is a fair point.
So, again, what is the distinction? Do we mark timeouts as `CONNECTION_LOST` and server returned errors, like file not found, or access denied/socket closed etc. as `DOWNLOAD_FAILED`. Do we then also need to make the distinction that for FAILED case the retry might not help, and for CONNECTION_LOST, you should try rettrying.


Lets wait comments from other reviewers.

I would like for example @lemrey @tokangas to reply. At least you guys have been working with fota_download library.

Reply:
I understand the point, but I'm not sure I understand the reasoning for the actual changes in the PR.

The new error cause `FOTA_DOWNLOAD_ERROR_CAUSE_CONNECTION_LOST` is used only in one case, if Download client returns ECONNRESET or ETIMEDOUT. What about for example the connection establishment?

Looking at where `FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED` is currently used, it's either for errors from the Download client or the DFU target library. Are the errors from the DFU target library related to the download? Maybe this is a wrong cause to use for those? Maybe at least part of these could be categorized differently.

Most (if not all) errors from the Download client can be categorized as download failures and can be transient in nature, I think. For example, DNS query, TCP connection establishment, TLS negotiation can all fail due to a transient connectivity, network or server issue.

To put it this the other way around, which one of the errors currently mapped to `FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED` is such that the application should for sure never retry? I'm not familiar enough with the DFU target library to say how each of the errors from the library should be handled.

Reply:
Thanks for your input. I see that its quite difficult to distinguish between CONNECTION_LOST and DOWNLOAD_FAILED. I think I drop changes related to that and put this changes on my ToDo-List. Once I have a clear idea of what has to be done and how, I will open another PR for that. Thanks again :)